src/main/scala/util/SelectDiplomacy.scala

```diff
@@ -0,0 +1,407 @@
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.rocket._
+import freechips.rocketchip.diplomacy.{LazyModuleImp, LazyModule, BaseNode}
+import freechips.rocketchip.diplomaticobjectmodel.HasLogicalTreeNode
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree._
+import freechips.rocketchip.diplomaticobjectmodel.model.{OMComponent, OMInterrupt}
+
+import chisel3.aop.{Aspect, Select}
+
+import firrtl.graph._
+
+import scala.collection.mutable
+
+import scala.collection.JavaConverters._
+
+trait GraphType
+object DOTGRAPH extends GraphType
+object GRAPHML extends GraphType
+
+case class RTreeData (val item: Any)
+
+
+object SelectDiplomacy {
+
+  val digraph = new MutableDiGraph[RTreeData]
+
+  def findNode(name:String): Any = {
+    None
+  }
+
+  /** Collects all Lazy Module Imp as a List.
+    */
+  def collectImpModules(): List[LazyModuleImp] = {
+    val edges = digraph.getEdgeMap
+    var lst: List[LazyModuleImp] = List[LazyModuleImp]()
+    edges.foreach { case (parent, children) =>
+      parent.item match {
+        case x: LazyModuleImp => lst = x :: lst
+        case _ => None
+      }
+    }
+    lst
+  }
+
+  /** Create a file with all Lazy Module Imp names.
+    */
+  def collectImpModules(targetDir: String, fname: String): String = {
+    val sb = new mutable.StringBuilder()
+    val edges = digraph.getEdgeMap
+    edges.foreach { case (parent, children) =>
+      parent.item match {
+        case x: LazyModuleImp => sb.append(s"${x.getClass.getName}.${x.name}\n")
+        case _ => 
+      }
+    }
+    writeFile(targetDir, fname, sb.toString)
+    sb.toString
+  }
+
+  /** Collects all BaseNodes as a List.
+    */
+  def collectBaseNodes(): List[BaseNode] = {
+
+    val vertices = digraph.getVertices
+    var lst: List[BaseNode] = List[BaseNode]()
+    vertices.foreach { case x =>
+      x.item match {
+        case x: BaseNode => lst = x :: lst
+        case _ => None
+      }
+    }
+    lst
+  }
+
+  /** Create a file with all Base Nodes names.
+    */
+  def collectBaseNodes(targetDir: String, fname: String): String = {
+    val sb = new mutable.StringBuilder()
+    val edges = digraph.getEdgeMap
+    edges.foreach { case (parent, children) =>
+      parent.item match {
+        case x: BaseNode => sb.append(s"${x.getClass.getName}.${x.name}\n")
+        case _ => 
+      }
+    }
+    writeFile(targetDir, fname, sb.toString)
+    sb.toString
+  }
+
+  /** Collects all Lazy Modules as a List.
+    */
+  def collectLazyModules(): List[LazyModule] = {
+    val edges = digraph.getEdgeMap
+    var lst: List[LazyModule] = List[LazyModule]()
+    edges.foreach { case (parent, children) =>
+      parent.item match {
+        case x: LazyModule => lst = x :: lst
+        case _ => None
+      }
+    }
+    lst
+  }
+
+  /** Create a file with all Lazy Module names.
+    */
+  def collectLazyModules(targetDir: String, fname: String): String = {
```

Comment:
Can this function reuse `def collectLazyModules(): List[LazyModule]` and then serialize/write the output?